### ISSUE TITLE
JMdict NG (next generation) progress and release date

### ISSUE THREAD

**Issue by parfait8566** [JMdictProject/JMdictIssues#136]
Previous two issues were closed.
In #116 early summer was put on the table, but I guess it might have been delayed. Again, I really don't want to put any pressure and I'm just curious about the progress.
One question I have is if it wouldn't be better to have some sort of "pitch accent box" in JMdictDB where the community can add or edit info in order to already cover a lot of entries when JMdict NG officially releases.
Side note, I think that the NHK accent dictionary also includes the pitch accent of inflected verbs and adjectives. Will this be included in JMdict?

**Comment by parfait8566** [comment]
Another question: pitch accent can change depending on POS. Sometimes JMdict has different POSs in the same sense. Would this be handled correctly?

**Comment by yamagoya** [comment]
Sorry, I missed the Aug 31 comment.

Regarding progress, I ran into some health issues this year that have got in the way.  Am trying to work around those.

Will this [NHK accent dictionary] be included in JMdict?  Not by me although if someone wants to extract and reformat the info there for inclusion in the database, that would be fine with me.  A look will probably need to be taken at potential copyright issues though.

Would this [pitch accent dependencies on POS] be handled correctly?  Currently no, there is no provision for different pitch accents for different parts-of-speech.

**Comment by stephenmk** [comment]
> A look will probably need to be taken at potential copyright issues though.

I'm not a lawyer, but my understanding is that [sweat of the brow](https://en.wikipedia.org/wiki/Sweat_of_the_brow) copyright protections don't exist in [Australia](https://www.mondaq.com/australia/intellectual-property/290668/can-a-database-be-protected-by-copyright), the US, and the UK. So large collections of facts such as telephone numbers, television broadcast schedules, or pitch accent locations wouldn't be copyrightable even if the collection of the data required substantial original research and labor. No idea what the legal situation is in Japan, though.

**Comment by JMdictProject** [comment]
I think the NG is quite a way off. Once the revised database is ready, there will be a complex task of changing over to it as well as keeping all the legacy systems working.

We haven't considered how the JMdictDB interface will look, but I do not favour introducing more boxes. The JEL language approach used for the information fields should be able to handle additional information such as pitch accents associated with readings. As for including information such as pitch accent from various sources, I expect that eventually some form of bulk update may be possible. Of course, the NHK ones are a possible source.

Re pitch accent and POS, we'll have to consider that issue eventually. It won't be simple.

**Comment by parfait8566** [comment]
Has there been progress on the JMdict NG release?

If not, can we consider adding now features/tags discussed in the past or are they strictly reserved for after NG comes out?
E.g. [gikun] kanji tag or selective [ateji]/[gikun] tag in #143 or police slang, [colk] (colloquial kana), etc. in #112

**Comment by yamagoya** [comment]
Yes, but it has been slow.  One of the holdups in deploying to Jim's site is that I have been trying to finish up a lot of loose ends related to the whole package, eg, the parsing/loading of all the xml files (jmdict, jmnedict, kanjidic).  I have proposed to Jim that I focus on finishing up only that changes he needs (he does not need to parse the xml files or load them into a database) and I'll finish the other stuff later.  Another slowdown is that the project is too big now to make changes without tests and writing them after the fact is very time consuming.

To provide a little more transparency, I will, within a few days, push the xml-ng branch code to Gitlab.  The reason I haven't done this before is not a desire for secrecy, but that I am a klutz when it come to Git and I find myself frequently redoing large parts of the Git history -- something easy to do when it is local, but very difficult after the code has been pushed to a public repository.

As for the remaining (edrdg-related) work, I think the major item is some issues with xrefs and their display.  There are a lot of little cleanup/fixup things as well.

**Comment by robinjmdict** [comment]
I'm very annoyed with myself because over two years ago I commented that I would share my ideas and proposals for JMdict NG, but I never got around to doing it and it completely slipped off my radar.

If it's not too late, I would still very much like to outline these proposals. I can do so tomorrow, either as a comment here or in a new issue. Given that this is a major update to JMdict, I think it's important that we don't end up regretting not including a feature that would have been very useful.

I sincerely apologise to yamagoya, as I appreciate that any additional changes mean more work for him and may further delay the release.

**Comment by yamagoya** [comment]
@robinjmdict, no promises, but I would like to see your suggestions.

**Comment by yamagoya** [comment]
A demo of the upcoming xml-ng version in its current state is available at:

  https://www.edrdg.org/jmdemo/

It looks pretty much like the old version, though it has some minor improvements (eg, try it on a small-screen device).  It is connected to a throw-away database so feel free to make whatever crazy submissions you wish.  You can verify the connected database at the bottom right of any page: it will say "DB=jmdemo" rather than "DB=jmdict".  Documentation for the user-facing changes is at:

 https://www.edrdg.org/jmdemo/web/doc/2025-09-update.html

(The date in the url was a little over-optimistic.)

If you want to see the XML for any entry, use the "View entry in alternate formats"
links at the bottom of the entry.

Feedback welcome and please keep in mind that you are not seeing the final version.

**Comment by yamagoya** [comment]
Pushed the xml-ng code to Gitlab:
  Code:  https://gitlab.com/yamagoya/jmdictdb/-/tree/xng
  Commits:  https://gitlab.com/yamagoya/jmdictdb/-/commits/xng?ref_type=heads

I will still probably only push additional changes intermittently in batches.

**Comment by parfait8566** [comment]
- I'm not sure if this is a bug, but the site seems to accept stuff like `[pa=0;1;2]` (to denote that there are multiple accepted patterns) but once submitted only one is saved (also ideally the pitch accent patterns should be ordered, e.g. if I write  `[pa=2;0]` pitch accent pattern 2 should be shown before pattern 0)
- I believe we should address pitch accent potentially changing by POS (e.g. 昨日 is `[pa=2]` as a noun but `[pa=0]` when used adverbially)
- some entries change pitch accent depending on the meaning (e.g. 具, クラブ, etc.)
- patterns can have important connotations and there should be ways to classify them as standard, traditional, young/modern, etc. (e.g. 彼氏)
- features in NHK like devoicing (e.g. shi in KE-shi-ki/景色) or nasalization are missing

I feel like it's important that JMdict's pitch accent system at the very least accepts multiple patterns and distinguishes between senses and POSs. I suspect e.g. Kansai-ben pitch accent is more complex than the standard as well but I'm not too familiar.

Also, have you looked at issue #143? It'd be very neat to be able to select kanji with the [ateji] tag or potential kanji [gikun] tag e.g. my idea was
- {図}抜ける[ateji] for 図抜ける
- {躊躇}う[gikun] for 躊躇う
- {梅雨}寒[gikun] for 梅雨寒
- {上衣}[gikun=うわぎ] for 上衣 (I'm aware that this particular example may be controversial as it's slightly old and nowadays we generally prefer separate entries for etymologically different readings but it still seems safe to have the option anyway even if it's rarely or never used)

**Comment by parfait8566** [comment]
> I'm very annoyed with myself because over two years ago I commented that I would share my ideas and proposals for JMdict NG, but I never got around to doing it and it completely slipped off my radar.

I believe @Marcusjmdict also had some proposals. Hope you don't mind the ping.

**Comment by robinjmdict** [comment]
@yamagoya My proposals for JMdict NG.


### Entry-wide cross references

Currently, x-refs can only be added at the sense level and they must point to a specific sense. It would be helpful to have x-refs that apply to a whole entry. For example, on [感ずる](https://www.edrdg.org/jmwsgi/entr.py?svc=jmdict&e=2325761), each sense has an x-ref to the corresponding sense in 感じる. This could be handled more cleanly with a single entry-wide x-ref pointing to the 感じる entry as a whole.

### Cross references in notes

When a sense note mention other Japanese terms, we often include x-refs to those terms. I think we should have the option of embedding these x-refs directly within the note text, so that the referenced terms themselves function as links. This would mean that users don't have to look for them at the bottom of the entry where they're mixed in with other x-refs for synonyms, related terms, unabbreviated forms, etc.

### Multiple languages in language source tags

When we want to indicate that a loanword is composed of multiple elements from different source languages (e.g. [ロールパン](https://www.edrdg.org/jmwsgi/entr.py?svc=jmdict&e=2219730), from English roll and Portuguese pão), it's currently necessary to use a separate [lsrc] tag for each constituent. Although we use the "partial" argument in these cases, many (most?) clients ignore it (e.g. jisho.org displays “From English roll; From Portuguese pão”). I think this information should be captured in a single tag. This would also allow us to mark the whole term as "wasei".

### Etymology note type

We have many sense-level notes explaining the etymology of terms. I suggest adding an "etymology" note type for the new entry-wide notes in JMdict NG. Support for this at the sense level might also be useful (See the ["JK" entry](https://www.edrdg.org/jmwsgi/entr.py?svc=jmdict&e=2328823) for an example).

### Entry-wide misc tags

Useful when [form], [dated] [poet], etc. apply to every sense of an entry. I think [yoji] and [on-mim] should *always* be entry-wide.

### Entry-wide dialect element

Currently only possibly to add at the sense level. Useful when the word itself is dialectal, rather than just specific senses.

### "OR" element for misc tags

In the kokugos, many terms are described as "derogatory or familiar", "dated or dialectal", etc. Currently, we can only do this with notes.

### Italics element

We have many sense notes that mention titles of books, films, etc. I think these titles should be italicised, as is standard practice in other dictionaries. It could also be used in etymology notes, e.g. ["from _corded_ + 天鵞絨"](https://www.edrdg.org/jmwsgi/entr.py?svc=jmdict&e=2082788).


I have a couple more suggestions that I'll post tomorrow.

**Comment by parfait8566** [comment]
I'm sneaking #167 in the thread just so that it doesn't get lost.



> ### "OR" element for misc tags
> In the kokugos, many terms are described as "derogatory or familiar", "dated or dialectal", etc. Currently, we can only do this with notes.

I think Marcus also proposed something similar in #146 (along with qualifiers e.g. [somewhat] formal), it'd be a very useful feature.

On the topic of cross-references, #39 (more precise cross-references) is also very much needed.

**Comment by yamagoya** [comment]
@parfait8566 
> I'm not sure if this is a bug, but the site seems to accept stuff like [pa=0;1;2] (to denote that there are multiple accepted patterns) but once submitted only one is saved (also ideally the pitch accent patterns should be ordered, e.g. if I write [pa=2;0] pitch accent pattern 2 should be shown before pattern 0)

Yes, it was a bug.  The input was parsed and stored correctly but only the first pa item was displayed.  It's been fixed.  Thanks for the heads up.  And yes, the pa items maintain their order.

**Comment by robinjmdict** [comment]
More suggestions:

### Glosses separated by part of speech

When glossing a term that has more than one part of speech, "the form of the [gloss] should agree with the first part-of-speech tag for the sense" (per our editorial policy). This is fine, but it means we often omit glosses that would be useful to learners and translators (e.g. adjective glosses for 形容動詞, 〜する glosses for adverbs). Sometimes we add an extra sense to accommodate these glosses, but this is a clunky workaround because it's really the same sense of the word.

I propose allowing multiple sets of glosses on a given sense, each corresponding to a particular part of speech. This is something that GG5 supports and uses extensively.

![Screenshot of GG5's entry for 自由](https://i.postimg.cc/v8qQMYdV/sdfgsdfgsdf.png)

This could be used not only for parts of speech like adj-no, adj-f and vs, but also particular forms in which a term commonly appears, e.g. 〜に, 〜で, 〜(と)した.

For example, [秘密裏](https://www.edrdg.org/jmwsgi/entr.py?svc=jmdict&e=2318842) is a noun but it's almost always used adverbially with に. Glossing it a noun would be too awkward so instead we provide adverbial glosses along with a "usu. adverbially as 〜に" note.

With this proposed new system, it could look like this:

> 秘密裏
> [n]
> 〜に in secrecy; behind closed doors


Obviously there would need to be a discussion about how this should be represented in XML. It would be a significant change to the existing structure but I believe it could be one of the most useful upgrades to JMdict.

### Gloss separators

For grouping together related glosses on a particular sense. [Wadoku does this (using ||)](https://www.wadoku.de/entry/view/6684713) and I think it works well.

**Comment by yamagoya** [comment]
Thanks for the suggestions, they seem like they would be useful but most involve substantial work with major changes to the database schema and downstream effects throughout the code down to the user interface and xml product files.

Some comments on some that might be doable ether before xml-ng release or shortly thereafter:

pitch accent types:
  In the case of multiple pa for a reading it does seem like something is needed to explain why/when each is applicable.  Would a note for each pa suffice for now?  A note could also describe PoS or sense restrictions.  The immediate problem is how to display the note since just appending it to the pa pattern might make too much of a jumble of the reading text.

Entry-wide cross references:
  This was planned for xml-ng but there are still some issues and it may be delayed until later.  I am still looking at it.

Cross references in notes:
  Can I get some examples?

Etymology note type:
  I think Jim can create an additional entry info tag for this.

"OR" element for misc tags
  How many tags like this are there?  What if there were simply additional Misc tags like "derogatory or familiar", "dated or dialectal"?

The other items seem sufficiently hard in either definition or implementation that they will need to be considered later in xml-ng follow on work.  Considering how long it has taken me so far, I really want to concentrate on getting the xml-ng work, even if not perfect, done, and want to avoid anything that will delay it further.  But certainly improvements don't have to stop with the xml-ng changes.

**Comment by parfait8566** [comment]
> ### pitch accent types:
> In the case of multiple pa for a reading it does seem like something is needed to explain why/when each is applicable. Would a note for each pa suffice for now? A note could also describe PoS or sense restrictions. The immediate problem is how to display the note since just appending it to the pa pattern might make too much of a jumble of the reading text.

I don't really like this approach.
彼氏, for example, has 2 senses: 1) boyfriend and 2) he; him. We'd have to specify in the same note that this entry has a more modern pitch accent 0 that applies only sense 1. This could get very ugly very quickly if the entry has a lot of senses or POSs.
The `reading[pa=]` format probably works for a strong majority of entries and is nice when all the senses follow the same patterns, but I think we still need at least
1. 2 registers (standard/official and modern)
2. per-sense [pa] overrides
3. POS qualifiers


Fiddling with AI chats I was given this as a viable approach:
```xml
<!ATTLIST re_pa 
    pa_type     CDATA #IMPLIED  <!-- Existing: "HL", etc. -->
    pos         CDATA #IMPLIED  <!-- Restriction: "n", "adv", "vs", etc. -->
    sense_index CDATA #IMPLIED  <!-- Restriction: "1", "2", "1,3" (CSV supported) -->
    status      CDATA #IMPLIED  <!-- Metadata: "modern", "standard", "dialect" -->
>
```
Entry 彼氏:
```xml
<entry>
  <ent_seq>1483140</ent_seq>
  <r_ele>
    <reb>かれし</reb>
    <!-- 
       Pattern 1 (KA-re-shi): 
       The standard/traditional accent.
       By default, applies to all senses unless overridden.
       status="std" helps apps label it "Standard" or "Traditional".
    -->
    <re_pa pa_type="std">1</re_pa>

    <!-- 
       Pattern 0 (ka-re-shi): 
       Restricted to Sense 1 (Boyfriend).
       Marked as "modern" so apps can warn the user or color-code it.
    -->
    <re_pa pa_type="modern" sense_index="1">0</re_pa>
  </r_ele>

  <!-- Sense 1: Index 1 -->
  <sense>
    <pos>&n;</pos>
    <gloss>boyfriend</gloss>
  </sense>

  <!-- Sense 2: Index 2 -->
  <sense>
    <pos>&pn;</pos>
    <gloss>he</gloss>
    <gloss>him</gloss>
  </sense>
</entry>
```
Entry 昨日 (only きのう for semplicity):
```xml
<entry>
  <ent_seq>1579260</ent_seq>
  <r_ele>
    <reb>きのう</reb>
    <re_inf>&gikun;</re_inf>
    
    <!-- 
      Pattern 2 (ki-NO-u): 
      Applies when used as a NOUN.
    -->
    <re_pa pos="n">2</re_pa>

    <!-- 
      Pattern 0 (ki-no-u): 
      Applies when used as an ADVERB.
    -->
    <re_pa pos="adv">0</re_pa>
  </r_ele>
  
  <sense>
    <!-- 
      This sense declares it can be EITHER 'n' or 'adv'. 
      The app checks the <re_pa> tags above to see if 
      specific rules exist for these tags. 
    -->
    <pos>&n;</pos>
    <pos>&adv;</pos>
    <gloss>yesterday</gloss>
  </sense>
</entry>
```

Let's imagine a generic word with three different senses and three different PAs.
PA 0 applies to all senses, PA 1 is also viable for sense 1 and sense 2 and PA 2 is the preferred pattern for sense 3.
Result:
```xml
<r_ele>
    <reb>reading</reb>
    
    <!-- 
       GROUP A: Sense 1 and 2 
       We explicitly link these pitches to 1 and 2.
       This prevents Sense 3 from accidentally "inheriting" Pitch 1.
    -->
    <re_pa sense_index="1,2">0</re_pa>
    <re_pa sense_index="1,2">1</re_pa>

    <!-- 
       GROUP B: Sense 3 
       We explicitly define the order for Sense 3.
       Notice we repeat '0' here. This is necessary because 
       it appears in a different priority position (2nd) 
       and is part of a different logical group.
    -->
    <re_pa sense_index="3">2</re_pa>
    <re_pa sense_index="3">0</re_pa>
</r_ele>
```

**Comment by parfait8566** [comment]
>### "OR" element for misc tags
>How many tags like this are there? What if there were simply additional Misc tags like "derogatory or familiar", "dated or dialectal"?

There are quite a few e.g. おっぱい would be [childish] or [colloquial], there are tons of [archaic] or [poetic] entries, etc. It's not very predictable so a flexible system would be much needed. Note that combined tags like [dated or dialectal] would need sub-options e.g. ksb (Kansai-ben).

>### Cross references in notes:
>Can I get some examples?

Entry [匆々](https://www.edrdg.org/jmwsgi/entr.py?svc=jmdict&q=2435210) has a note that says "letters ending with this start with 前略, 冠省, etc." and entry [ただただ](https://www.edrdg.org/jmwsgi/entr.py?svc=jmdict&q=1596990) one that says "emphatic form of ただ". Latter example is a bit tricky because I'd rather like some sort of [emphatic] tag and/or x-ref that goes "emphatic version of X" rather than resorting to a note.
I think we should in general have a x-ref system within the glosses. There are entries that require very Japanese concepts and when glossed we either use the Japanese term itself (which learners could not know) or an awkward translation. Being able to "hyperlink" the terms would make using the dictionary much smoother.

**Comment by parfait8566** [comment]
Regarding the #143 issue, would this XML implementation work?
```xml
<!ELEMENT k_struct (#PCDATA | seg)*>
<!ELEMENT seg (#PCDATA)>
<!ATTLIST seg 
    type (ateji | gikun) #REQUIRED
    reading CDATA #IMPLIED>
```

For {躊躇}う[gikun]:
```xml
<entry>
  <ent_seq>1573390</ent_seq>
  <k_ele>
    <keb>躊躇う</keb>
    <k_struct><seg type="gikun">躊躇</seg>う</k_struct>
    <ke_pri>ichi1</ke_pri>
  </k_ele>
  <r_ele>
    <reb>ためらう</reb>
    <re_pri>ichi1</re_pri>
  </r_ele>
  <!-- ... sense ... -->
</entry>
```
For {梅雨}寒[gikun]:
```xml
<k_ele>
  <keb>梅雨寒</keb>
  <k_struct><seg type="gikun">梅雨</seg>寒</k_struct>
</k_ele>
```
For {図}抜ける[ateji]:
```xml
<k_ele>
  <keb>図抜ける</keb>
  <k_struct><seg type="ateji">図</seg>抜ける</k_struct>
</k_ele>
```
For {上衣}[gikun=うわぎ]:
```xml
<k_ele>
  <keb>上衣</keb>
  <!-- Explicitly binds this gikun tag to the reading 'uwagi' -->
  <k_struct><seg type="gikun" reading="うわぎ">上衣</seg></k_struct>
</k_ele>
<r_ele>
  <reb>うわぎ</reb> <!-- Matches the tag -->
</r_ele>
<r_ele>
  <reb>じょうい</reb> <!-- Does NOT match, so this remains a standard reading -->
</r_ele>
```

**Comment by yamagoya** [comment]
[Edited to be a little less terse.] 

Those may be feasible XML (haven't looked closely yet) but XML is only one part of the implementation.  The XML is produced from the database thus has to be representable in the database.  The tables for entries are defined in:

  https://gitlab.com/yamagoya/jmdictdb/-/blob/6790ebba3b62d18d011e8c94e0ecebb85df3a1bc/db/entrobjs.sql

The "pa" table is at line 320 for example.  The tables more or less correspond to XML elements with containment implemented by a foreign key to the parent element/table.

And a syntax for describing the features has to be modified/added in the JEL parser (parses the text from edit page reading, kanji, and sense boxes):

  https://gitlab.com/yamagoya/jmdictdb/-/blob/6790ebba3b62d18d011e8c94e0ecebb85df3a1bc/jmdictdb/jelparse.y

The "[pa={hl-string}>]" syntax is recognized at line 218 and the [pa={number}] syntax at line 304, both in the "tagitem" production.

And (probably not lastly but minimally) the template for displaying entries in the JMdictDB "entr.py" web page has to be modified to sensibly display the feature.  This is likely the easiest part.

Whatever proposal is adopted has to be implementable in all three of those places.  If you try getting an AI to do this I'm definitely interested in seeing the results.

**Comment by parfait8566** [comment]
Table:
```sql
CREATE TABLE pa (       -- Pitch accents for readings.
    entr INT NOT NULL,
    rdng SMALLINT NOT NULL,
      FOREIGN KEY (entr,rdng) REFERENCES rdng(entr,rdng)
        ON DELETE CASCADE ON UPDATE CASCADE,
    
    -- The priority order. Crucial for the "Sense 3 prefers pattern 2" scenario.
    ord SMALLINT NOT NULL DEFAULT(1),
    
    -- The raw pitch data (High/Low bits)
    pitch BIT VARYING(4096),
    
    -- Disambiguates accents on the last mora (High-Low vs High-flat)
    acc BOOL NULL
      CHECK (substring (pitch FROM length(pitch) FOR 1)='0'
             OR acc IS NOT NULL),
             
    -- Display format: 'a' (ascii/text like HL) or 'n' (numeric like 0, 1)
    disp CHAR(1) NULL
      CHECK (disp IN ('a','n')),

    -- NEW COLUMNS (To support your XML Proposal) ----------------

    -- 1. SENSE RESTRICTION
    -- Maps to XML attribute: sense_index="1,2"
    -- Stored as a Postgres Array of integers for efficient querying.
    -- NULL = Applies to all senses.
    senses SMALLINT[] NULL, 

    -- 2. POS RESTRICTION
    -- Maps to XML attribute: pos="n" or pos="n,vs"
    -- Stored as text. NULL = Applies to all POS.
    -- Increased size to 64 to handle combined POS tags safely.
    pos VARCHAR(64) NULL,

    -- 3. STATUS / REGISTER
    -- Maps to XML attribute: status="modern" or status="std"
    status VARCHAR(32) NULL
    --------------------------------------------------------------
);
```

The proposed JEL syntax is `[pa=PATTERN:SENSES:POS:STATUS]`
Use a separator (e.g., `:`) to delimit the fields.
Empty fields inherit defaults.
Examples:
Standard numeric: `[pa=0]` (Unchanged behavior)
Sense restricted: `[pa=0:1]` (Pattern 0, Sense 1 only)
Sense grouping: `[pa=0:1,2]` (Pattern 0, Senses 1 and 2)
POS restricted: `[pa=2::n]` (Pattern 2, any sense, Noun only)
Modern flag: `[pa=0:::modern]` (Pattern 0, all senses, all POS, Modern label)
Implementation:
```py
| TEXT EQL TEXT         
                { 
                if p[1] == 'pa':
                    # Split the value by colon to extract metadata
                    # Input format: [pa=pattern:senses:pos:status]
                    # Example: "0:1,2:n:modern"
                    parts = p[3].split(':')
                    
                    pattern_str = parts[0]
                    senses = None
                    pos = None
                    status = None

                    # Extract extended fields if they exist
                    if len(parts) > 1 and parts[1]:
                        # "1,3" -> "1,3" (kept as string here, converted to list/array later)
                        senses = parts[1] 
                    if len(parts) > 2 and parts[2]:
                        pos = parts[2]
                    if len(parts) > 3 and parts[3]:
                        status = parts[3]

                    # VALIDATION
                    # 1. Check if it's a valid pitch string (L/H) OR a Number
                    # The original code only allowed [LH]+ here. We now allow digits 
                    # because "pa=0:1" will be parsed as TEXT, not NUMBER.
                    if not (re.match(r'^[LH]+$', pattern_str, re.I) or re.match(r'^\d+$', pattern_str)):
                        perror ('The "pa" tag value must start with an accent number or pitch string (eg "0" or "LHHL")')

                    # 2. Check for "clean" numeric (without metadata).
                    # If p[3] is JUST a number (e.g. "2"), it technically might fall into 
                    # the "TEXT EQL jrefs" block depending on lexer priority. 
                    # But if it lands here, we treat it as valid.
                    
                    # Construct the object to return. 
                    # We append a dictionary or tuple with the extra metadata.
                    # Original: p[0] = [['pa', p[3], 'pitch']]
                    
                    p[0] = [['pa', pattern_str, 'pitch', {'senses': senses, 'pos': pos, 'status': status}]]
                    
                else: 
                    p[0] = [tag_eql_text (p, p[1], p[3])] 
                }
```
```py
# Pseudo-code for the DB insertion logic
for tag in taglist:
    if tag[0] == 'pa':
        pitch_val = tag[1]
        meta = tag[3] # The dict we added: {'senses': '1,2', ...}
        
        # Prepare SQL values
        senses_arr = None
        if meta['senses']:
            # Convert CSV string "1,2" to Postgres Array format "{1,2}"
            senses_arr = "{" + meta['senses'] + "}"
        
        # INSERT INTO pa (entr, rdng, pitch, senses, pos, status) ...
```

So 昨日 would look like `[pa=2::n][pa=0::adv]` and `[pa=0:1::modern][pa=1:1::std][pa=1:2::std]` (assuming modern PA should be first)

**Comment by yamagoya** [comment]
Interesting.  But it confirms what I feared.  I have been playing with AI a bit and it has been very useful creating some localized code that was not dependent on on a broad knowledge on the project architecture.  But I think the JMdictDB project is probably a good example of a very AI-unfriendly one.  A big problem is that it is not highly localized: to alter in a significant way most parts of the code requires general understanding of a wide swath of much the code as well as some overarching design principles that apply to everything.  So unless an AI is given all this context, it is unlikely to generate compatible or even close to working solutions. 

A specific problem immediately obvious in the "pa" table is the lack of any referential integrity (ie, use of foreign keys).  One could put 999999 or -15 as a sense number in pa.sense with no complaints from the database.  Compare for example with the xref table where the reference to the target's kanji or reading in enforced by a foreign key.

The same issue applies to the pa.pos column.  Again, compare to a PoS in a sense where multiple PoS are individual rows in a "pos" table and validity is enforced by a foreign key referencing the PoS tag table ('kwpos").

The JEL parser code is also not viable, again probably because the AI did not have enough context.  It looks like it assumed that something called TEXT was an arbitrary text string but it is not; it is the name of a "token" received from the lexer, jellex.py.  What the parser will receive if one gives it the text string "えいが[pa=0:1,2]" is the following sequence of tokens:

```
$ tools/dbg-parser.py lex
Enter text (can be multi-line), ^D to parse, ^C to exit
えいが[pa=0:1,2]
LexToken(RTEXT,'えいが',1,0)
LexToken(BRKTL,'[',1,3)
LexToken(TEXT,'pa',1,4)
LexToken(EQL,'=',1,6)
LexToken(NUMBER,'0',1,7)
LexToken(COLON,':',1,8)
LexToken(NUMBER,'1',1,9)
LexToken(COMMA,',',1,10)
LexToken(NUMBER,'2',1,11)
LexToken(BRKTR,']',1,12)
done
```

As you can see, the colons (and other bits) will be received as separate tokens, not as characters in a string.

If you're willing to share, I'd be interested to know what AI you used, if it was free or paid, and which files you gave it for context?

Finally, a very old to-do list item for me has been to write some sort of "architecture guide" -- a high level technical description of how the various parts of the system work together and what is implemented where.  Such a guide would also be immensely helpful to an AI as well.  But it will have to wait until after the xml-ng stuff is finished.

**Comment by parfait8566** [comment]
> Interesting. But it confirms what I feared. I have been playing with AI a bit and it has been very useful creating some localized code that was not dependent on on a broad knowledge on the project architecture. But I think the JMdictDB project is probably a good example of a very AI-unfriendly one. A big problem is that it is not highly localized: to alter in a significant way most parts of the code requires general understanding of a wide swath of much the code as well as some overarching design principles that apply to everything. So unless an AI is given all this context, it is unlikely to generate compatible or even close to working solutions.

Maybe you could try using Cursor (free). Someone with a good understanding of the code would probably fare much better than me.
You can add a folder to the workspace, tag certain files (e.g. `@db/entrobjs.sql`) and help it steer where you want with the necessary context.

This is what I got:

---

# Enhanced Pitch Accent System: Register, Sense Restrictions, and POS Qualifiers

## Overview

Adds support for:
1. Two registers (standard/official and modern)
2. Per-sense pitch accent overrides
3. POS qualifiers for pitch accents

Uses proper foreign keys for referential integrity and token-based parsing compatible with the JEL lexer.

---

## 1. Database Schema Changes (`db/entrobjs.sql`)

### 1.1 Added `reg` field to `pa` table

Added a register field to distinguish standard/traditional from modern pitch accents:

```sql
CREATE TABLE pa (       -- Pitch accents for readings.
    entr INT NOT NULL,
    rdng SMALLINT NOT NULL,
      FOREIGN KEY (entr,rdng) REFERENCES rdng(entr,rdng)
        ON DELETE CASCADE ON UPDATE CASCADE,
    ord SMALLINT NOT NULL DEFAULT(1),
    pitch BIT VARYING(4096),
    acc BOOL NULL
      CHECK (substring (pitch FROM length(pitch) FOR 1)='0'
             OR acc IS NOT NULL),
    disp CHAR(1) NULL
      CHECK (disp IN ('a','n')),
    reg CHAR(1) NULL
      CHECK (reg IN ('s','m')),
      -- Register/type of pitch accent:
      --   's': standard/traditional (official)
      --   'm': modern
      --   NULL: default (no specific register specified)
      PRIMARY KEY (entr,rdng,ord));
```

**Changes:**
- Added `reg CHAR(1) NULL` with CHECK constraint allowing `'s'` (standard) or `'m'` (modern)
- Added `PRIMARY KEY (entr,rdng,ord)` to support junction tables

### 1.2 Created `pa_sens` junction table

Links pitch accents to specific senses with foreign key validation:

```sql
CREATE TABLE pa_sens (
      -- Links pitch accents to specific senses.  If a pa record has
      -- entries in this table, it applies only to those senses.
      -- If no entries exist, the pa applies to all senses.
    entr INT NOT NULL,
    rdng SMALLINT NOT NULL,
    pa_ord SMALLINT NOT NULL,
      FOREIGN KEY (entr,rdng,pa_ord) REFERENCES pa(entr,rdng,ord)
        ON DELETE CASCADE ON UPDATE CASCADE,
    sens SMALLINT NOT NULL,
      FOREIGN KEY (entr,sens) REFERENCES sens(entr,sens)
        ON DELETE CASCADE ON UPDATE CASCADE,
      PRIMARY KEY (entr,rdng,pa_ord,sens));
```

**Features:**
- Foreign key to `pa(entr,rdng,ord)` ensures valid pitch accent references
- Foreign key to `sens(entr,sens)` ensures valid sense numbers (prevents invalid values like 999999)
- If no entries exist in this table, the PA applies to all senses

### 1.3 Created `pa_pos` junction table

Links pitch accents to specific POS tags with foreign key validation:

```sql
CREATE TABLE pa_pos (
      -- Links pitch accents to specific part-of-speech tags.  If a pa
      -- record has entries in this table, it applies only when the
      -- sense has one of those POS tags.  If no entries exist, the pa
      -- applies regardless of POS.
    entr INT NOT NULL,
    rdng SMALLINT NOT NULL,
    pa_ord SMALLINT NOT NULL,
      FOREIGN KEY (entr,rdng,pa_ord) REFERENCES pa(entr,rdng,ord)
        ON DELETE CASCADE ON UPDATE CASCADE,
    kw SMALLINT NOT NULL
      REFERENCES kwpos(id) ON DELETE CASCADE ON UPDATE CASCADE,
      PRIMARY KEY (entr,rdng,pa_ord,kw));
```

**Features:**
- Foreign key to `kwpos(id)` ensures valid POS keywords (prevents invalid POS values)
- If no entries exist, the PA applies regardless of POS

---

## 2. Object Model Changes (`jmdictdb/objects.py`)

### 2.1 Updated `Pa` class

Extended the `Pa` class to include the new fields:

```python
class Pa (DbRow):
    def __init__ (s, entr=None, rdng=None, ord=None, pitch=None, acc=None, disp=None, reg=None,
                     _sens=None, _pos=None):
        DbRow.__init__(s, ( entr,  rdng,  ord,  pitch,  acc,  disp,  reg),
                          ('entr','rdng','ord','pitch','acc','disp','reg'))
        s._sens = _sens or []  # List of sense numbers this PA applies to (if empty, applies to all)
        s._pos  = _pos  or []  # List of POS kw id numbers this PA applies to (if empty, applies to all)
```

**Changes:**
- Added `reg` parameter to `__init__` and included it in `DbRow` initialization
- Added `_sens` list (temporary, for sense restrictions during parsing)
- Added `_pos` list (temporary, for POS restrictions during parsing)
- Follows the existing pattern where `_`-prefixed attributes are temporary lists not stored directly in the database

---

## 3. JEL Parser Changes (`jmdictdb/jelparse.py`)

### 3.1 Extended `pa` tag syntax

The parser now supports comma-separated attributes after the pitch accent value. All productions work with the token-based lexer (using `TEXT`, `NUMBER`, `COMMA`, `EQL` tokens).

#### 3.1.1 New productions for pitch strings with attributes

```python
def p_tagitem_5a(p):
    '''tagitem : TEXT EQL TEXT pa_attrs'''
    if p[1] == 'pa':
          # Only letters can occur in the pitch pattern or we
          #  wouldnt be in this production.
        if not re.match (r'[LH]+', p[3], re.I):
            perror (p, 'The "pa" tag allows only an accent number '
                    'or pitch string (eg "LHHL")')
        reg, sens_list, pos_list = p[4]
        p[0] = [['pa', p[3], 'pitch', reg, sens_list, pos_list]]
    else: perror (p, "Attributes only allowed with 'pa' tag")
```

Handles: `[pa=LHHL,modern]` or `[pa=LHHL,sens=1,2]`

#### 3.1.2 New production for numeric accent values with attributes

```python
def p_tagitem_5b(p):
    '''tagitem : TEXT EQL NUMBER pa_attrs'''
    if p[1] == 'pa':
        # Numeric accent value with attributes: pa=0,sens=1,2
        # Note: This is a simpler path than going through jrefs
        reg, sens_list, pos_list = p[4]
        p[0] = [['pa', int(p[3]), 'accent', reg, sens_list, pos_list]]
    else: perror (p, "Attributes only allowed with 'pa' tag")
```

Handles: `[pa=0,modern]` or `[pa=0,sens=1,2]`

#### 3.1.3 Updated existing production for backward compatibility

```python
def p_tagitem_5(p):
    '''tagitem : TEXT EQL TEXT'''
    if p[1] == 'pa':
          # Only letters can occur in the pitch pattern or we
          #  wouldnt be in this production.
        if not re.match (r'[LH]+', p[3], re.I):
            perror ('The "pa" tag allows only an accent number '
                    'or pitch string (eg "LHHL")')
        p[0] = [['pa', p[3], 'pitch', None, None, None]]  # Added None, None, None for backward compatibility
    else: p[0] = [tag_eql_text (p, p[1], p[3])]
```

Maintains backward compatibility: `[pa=LHHL]` still works.

### 3.2 Attribute parsing productions

#### 3.2.1 Attribute list parsing

```python
def p_pa_attrs_1(p):
    '''pa_attrs : COMMA pa_attr'''
    p[0] = p[2]

def p_pa_attrs_2(p):
    '''pa_attrs : COMMA pa_attr COMMA pa_attrs'''
    reg1, sens1, pos1 = p[2]
    reg2, sens2, pos2 = p[4]
    # Merge attributes, later ones override earlier ones
    reg = reg2 if reg2 is not None else reg1
    sens_list = sens2 if sens2 is not None else sens1
    pos_list = pos2 if pos2 is not None else pos1
    p[0] = (reg, sens_list, pos_list)
```

Handles comma-separated attributes like `,modern,sens=1,2,pos=n`

#### 3.2.2 Register attribute (shorthand)

```python
def p_pa_attr_1(p):
    '''pa_attr : TEXT'''
    # Shorthand for register: 'modern' or 'std'
    if p[1] in ('modern', 'std'):
        reg = 'm' if p[1] == 'modern' else 's'
        p[0] = (reg, None, None)
    else:
        perror (p, "Unknown pa attribute: '%s'. Use 'modern', 'std', 'sens=...', or 'pos=...'" % p[1])
```

Handles: `[pa=0,modern]` or `[pa=0,std]`

#### 3.2.3 Register attribute (explicit)

```python
def p_pa_attr_2(p):
    '''pa_attr : TEXT EQL TEXT'''
    # Explicit attribute: reg=modern, pos=n (single TEXT value)
    if p[1] == 'reg':
        if p[3] == 'modern':
            p[0] = ('m', None, None)
        elif p[3] == 'std':
            p[0] = ('s', None, None)
        else:
            perror (p, "Invalid register value '%s'. Use 'modern' or 'std'" % p[3])
    elif p[1] == 'pos':
        # Single POS keyword
        KW = jdb.KW
        pos_kw = KW.POS.get(p[3])
        if not pos_kw:
            perror (p, "Unknown POS keyword: '%s'" % p[3])
        p[0] = (None, None, [pos_kw.id])
    else:
        perror (p, "Unknown pa attribute: '%s'. Use 'reg' or 'pos'" % p[1])
```

Handles: `[pa=0,reg=modern]` or `[pa=0,pos=n]`

#### 3.2.4 Sense restrictions (single)

```python
def p_pa_attr_3(p):
    '''pa_attr : TEXT EQL NUMBER'''
    # Single sense number: sens=1
    if p[1] == 'sens':
        n = int(p[3])
        if n < 1 or n > 99:
            perror (p, "Invalid sense number: %d (must be 1-99)" % n)
        p[0] = (None, [n], None)
    else:
        perror (p, "Unknown pa attribute: '%s'. Use 'sens' for numeric values" % p[1])
```

Handles: `[pa=0,sens=1]`

#### 3.2.5 Sense restrictions (multiple)

```python
def p_pa_attr_4(p):
    '''pa_attr : TEXT EQL sens_nums'''
    # Comma-separated sense numbers: sens=1,2,3
    if p[1] == 'sens':
        p[0] = (None, p[3], None)
    else:
        perror (p, "Unknown pa attribute: '%s'. Use 'sens' for sense numbers" % p[1])

def p_sens_nums_1(p):
    '''sens_nums : NUMBER'''
    n = int(p[1])
    if n < 1 or n > 99:
        perror (p, "Invalid sense number: %d (must be 1-99)" % n)
    p[0] = [n]

def p_sens_nums_2(p):
    '''sens_nums : NUMBER COMMA sens_nums'''
    n = int(p[1])
    if n < 1 or n > 99:
        perror (p, "Invalid sense number: %d (must be 1-99)" % n)
    p[0] = [n] + p[3]
```

Handles: `[pa=0,sens=1,2,3]` - properly tokenized as `NUMBER COMMA NUMBER COMMA NUMBER`

#### 3.2.6 POS restrictions (multiple)

```python
def p_pa_attr_5(p):
    '''pa_attr : TEXT EQL pos_list'''
    # Comma-separated POS keywords: pos=n,adv
    if p[1] == 'pos':
        p[0] = (None, None, p[3])
    else:
        perror (p, "Unknown pa attribute: '%s'. Use 'pos' for POS tags" % p[1])

def p_pos_list_1(p):
    '''pos_list : TEXT'''
    KW = jdb.KW
    pos_kw = KW.POS.get(p[1])
    if not pos_kw:
        perror (p, "Unknown POS keyword: '%s'" % p[1])
    p[0] = [pos_kw.id]

def p_pos_list_2(p):
    '''pos_list : TEXT COMMA pos_list'''
    KW = jdb.KW
    pos_kw = KW.POS.get(p[1])
    if not pos_kw:
        perror (p, "Unknown POS keyword: '%s'" % p[1])
    p[0] = [pos_kw.id] + p[3]
```

Handles: `[pa=0,pos=n,adv]` - properly tokenized as `TEXT COMMA TEXT`

### 3.3 Updated `bld_rdng()` function

Modified to process the new attribute format:

```python
elif typ == 'pa':
   # Format: ['pa', value, type, reg, sens_list, pos_list]
   if len(t) >= 3:
       pa_value, pa_type = t[1], t[2]
       reg = t[3] if len(t) > 3 else None
       sens_list = t[4] if len(t) > 4 else None
       pos_list = t[5] if len(t) > 5 else None
   else:
       # Old format for backward compatibility
       pa_value, pa_type = t[1], t[2]
       reg, sens_list, pos_list = None, None, None
   try: pa = make_pa_obj (pa_value, pa_type, r.txt)
   except PaError as e: errs.append (str (e))
   except ParseError as e: errs.append (str (e))
   else:
       pa.reg = reg
       if sens_list: pa._sens = sens_list
       if pos_list: pa._pos = pos_list
       r._pa.append (pa)
```

**Changes:**
- Extracts `reg`, `sens_list`, and `pos_list` from the parsed tag list
- Maintains backward compatibility with old format (2-element list)
- Sets attributes on the `Pa` object for later processing

### 3.4 Updated `p_tagitem_11a` for jrefs with attributes

Also updated the production that handles `pa` tags via the `jrefs` production:

```python
def p_tagitem_11a(p):
    '''tagitem : TEXT EQL jrefs pa_attrs'''
    # Handle pa=number,attr1,attr2 syntax
    tag = p[1]
    if tag != 'pa':
        perror (p, "Attributes only allowed with 'pa' tag")
    # For pa tags, we only expect a single jref (accent number)
    if len(p[3]) != 1:
        perror (p, 'The "pa" tag allows only a single accent number')
    jpair, slist, seq, corpus = p[3][0]
    if corpus or slist:
        perror (p, 'The "pa" tag allows only an accent number '
                'or pitch string (eg "LHHL")')
    reg, sens_list, pos_list = p[4]
    p[0] = [['pa', seq, 'accent', reg, sens_list, pos_list]]
```

---

## 4. Syntax Examples

### 4.1 Basic usage (backward compatible)

```jel
かれし[pa=1]
かれし[pa=LHHL]
```

### 4.2 Register specification

```jel
# Shorthand
かれし[pa=1,std]        # Standard/traditional accent
かれし[pa=0,modern]     # Modern accent

# Explicit
かれし[pa=1,reg=std]
かれし[pa=0,reg=modern]
```

### 4.3 Sense restrictions

```jel
# Single sense
かれし[pa=1,sens=1]           # Pattern 1 for sense 1 only

# Multiple senses
かれし[pa=0,sens=1,2]         # Pattern 0 for senses 1 and 2
かれし[pa=1,sens=1,2,3]       # Pattern 1 for senses 1, 2, and 3
```

### 4.4 POS restrictions

```jel
# Single POS
きのう[pa=2,pos=n]            # Pattern 2 when used as noun

# Multiple POS
きのう[pa=2,pos=n,adv]        # Pattern 2 when used as noun or adverb
```

### 4.5 Combined attributes

```jel
# Register + sense restrictions
かれし[pa=1,std,sens=1,2]     # Standard pattern 1 for senses 1 and 2
かれし[pa=0,modern,sens=1]    # Modern pattern 0 for sense 1 only

# Register + POS restrictions
きのう[pa=2,std,pos=n]        # Standard pattern 2 when used as noun

# Sense + POS restrictions
きのう[pa=2,sens=1,pos=n]     # Pattern 2 for sense 1 when used as noun

# All combined
かれし[pa=0,modern,sens=1,pos=n]  # Modern pattern 0 for sense 1 when used as noun
```

### 4.6 Complex examples

```jel
# 彼氏 example: standard accent for all senses, modern accent for sense 1 only
かれし[pa=1,std][pa=0,modern,sens=1]

# 昨日 example: different accents for different POS
きのう[pa=2,pos=n][pa=0,pos=adv]

# Multiple patterns with different restrictions
reading[pa=0,sens=1,2][pa=1,sens=1,2][pa=2,sens=3][pa=0,sens=3]
```

---

## 5. Token-Based Parsing Verification

The implementation correctly handles tokenized input from the lexer. For example, `[pa=0,sens=1,2]` is tokenized as:

```
BRKTL('[')
TEXT('pa')
EQL('=')
NUMBER('0')
COMMA(',')
TEXT('sens')
EQL('=')
NUMBER('1')
COMMA(',')
NUMBER('2')
BRKTR(']')
```

The parser productions match these tokens correctly:
- `p_tagitem_5b`: `TEXT('pa') EQL('=') NUMBER('0') pa_attrs`
- `p_pa_attrs_1`: `COMMA(',') pa_attr`
- `p_pa_attr_4`: `TEXT('sens') EQL('=') sens_nums`
- `p_sens_nums_2`: `NUMBER('1') COMMA(',') sens_nums` → `NUMBER('2')`

---

## 6. Design Decisions

### 6.1 Foreign key integrity

- Uses junction tables (`pa_sens`, `pa_pos`) instead of arrays or VARCHAR columns
- Foreign keys ensure:
  - Sense numbers reference valid `sens(entr,sens)` rows
  - POS keywords reference valid `kwpos(id)` rows
  - Invalid values (e.g., sense 999999) are rejected by the database

### 6.2 Backward compatibility

- Old syntax `[pa=0]` and `[pa=LHHL]` still works
- New attributes are optional
- Existing entries continue to function

### 6.3 Token-based parsing

- All productions use actual lexer tokens (`TEXT`, `NUMBER`, `COMMA`, `EQL`)
- No string splitting or regex parsing of attribute values
- Compatible with the existing JEL lexer architecture

---

## 7. Next Steps

1. Create database migration script to add the new fields and tables to existing databases
2. Update XML export code (`jmxml.py`) to output the new attributes
3. Update database save/update code to handle the junction tables (`pa_sens` and `pa_pos`) when saving `Pa` objects
4. Update the web UI to support editing the new attributes
5. Add tests for the new syntax

---

This implementation provides a robust foundation for the enhanced pitch accent system while maintaining compatibility with existing code and data.

**Comment by yamagoya** [comment]
That's pretty impressive.  I've been using Aider with a few AIs, mainly Deepseek, but I haven't tried feeding it the full project because I have a rather low bandwidth internet connection.  A few comments:

1. I would still like to know what AI you used, whether it was free or paid, and what files you tagged.

2. If you have access to a Linux or Unix machine, would you consider setting up a development environment so you can modify and run the code locally?  This would allow you to wring out some of the inevitable problems with the generated code, even if you are not a programmer, by telling the AI what went wrong.  And of course if you know some Python or SQL, the process will be that much more efficient.  There is a **Development Guide** at: 
  https://www.edrdg.org/jmdemo/web/doc/index.html
Looking at it now, it needs some major rewriting but hopefully it's usable.  The **Install Guide** there is also relevant for needed software and database server.   I would be happy to help if you run into difficulties or have questions.  All the Python tools can (I think) be installed locally and the only "big" piece of software that may need system-wide installation is the Postgresql database server.  You don't need full blown web server for development work, the code comes with a small local one that requires no setup.

3. Could we move this discussion to a Gitlab issue, adjacent to the actual code.  This level of code discussion is out of scope for "JMdict NG (next generation) progress and release date" and for these Github issues in general ("JMdict Japanese dictionary - lexicographic, etc. issues management") although Jim can correct me if I wrong on the latter point.  I suggest copying your last post to a new Gitlab issue at:
  https://gitlab.com/yamagoya/jmdictdb/-/issues
and editing it here to replace all the text with a URL to the Gitlab issue.

4. I am very interested in pursuing this but right now I really need to finish the xml-ng work, and it is already apparent that these changes, even with AI help, are going to be too time consuming.  (Don't forget that documentation and test coverage will also be needed before inclusion.)  And, just to be clear, I can not make any promises about whether the code changes will ultimately be included in JMdictDB, even if they are working; that's a standard hazard with volunteer open source work. 

5. At some point Jim will need to weigh in on the changes but that can happen after the changes are known to be implementable and the XML and JEL syntax changes are firm.

6. Lastly but not leastly, thanks for your efforts on this!

**Comment by parfait8566** [comment]
>4. I am very interested in pursuing this but right now I really need to finish the xml-ng work, and it is already apparent that these changes, even with AI help, are going to be too time consuming. (Don't forget that documentation and test coverage will also be needed before inclusion.)

I feel strongly that if pitch accent support is implemented it should already come with *at least* sense and POS restriction features.
How would a "pa note" work for 彼氏? It'd need to say "pa=1, only for sense 1 pa=0 is also common and considered more modern". Apps/sites are going to have trouble showing this information in a sensible and intuitive way for users and would probably need to resort to e.g. custom regex (that invariably fails in edge cases). It's also going to get extremely messy when the entry has a lot of senses that don't follow the same pattern (e.g. 一杯) and when the restriction features are implemented later a lot of work is going to be needed to re-edit the problematic entries.
In my opinion it's best to delay the PA system altogether rather than have it in a form that's in all likelihood gonna cause a lot of headaches down the line.

>1. I would still like to know what AI you used, whether it was free or paid, and what files you tagged.

I used Cursor (basically an IDE with AI models built-in). It has a free plan (which is the one I used) but also a subscription for more usage and better models. I added the repo to the workspace and told it to look at the files `db/entrobjs.sql` and` jmdictdb/jelparse.py` (and it seems to have automatically touched other files like `jmdictdb/objects.py`). Keep in mind that even if you add the entire repo the context is still limited so it's not gonna actually properly read all the files.

I don't mind trying to actually run the code myself and moving the discussion to Gitlab, though I'd prefer if the repo was in Github as I wouldn't need to make an account (and it's way more popular/other people are significantly more likely to contribute).

**Comment by yamagoya** [comment]
> In my opinion it's best to delay the PA system altogether rather than have it in a form that's in all likelihood gonna cause a lot of headaches down the line.

I am fine with that if Jim @JMdictProject agrees and there are no vehement disagreements.

Re discussion location, I think discussing the specification / user-facing aspects of the PA system is fine here (tho in a new issue please), but detailed code and implementation discussions need to go where the code is because they will constitute part of the history of the code.

**Comment by parfait8566** [comment]
Yeah, I meant having the code/repo here on Github would be handy as it's just significantly more popular and therefore easier for other contributors to jump in and help. Anyway, soon enough I am gonna create a Gitlab account and create an appropriate issue there (though moving to Github down the line would probably be a plus in terms of development manpower).

**Comment by yamagoya** [comment]
It's easy enough to move the code repo, the problem is the issues.  Github AFAIK does not provide a way of easily importing issues.  I gather there are 3rd party tools but I'm doubtful they are as easy as a button click.  Part of the reason for putting the code on Gitlab was that it was easier to import the issues to there AND maintain the same issue numbers they had in the old in-house Roundup system.  The latter point was important because a lot of the early commits and my own notes referred to those issue numbers.  It turns out I haven't made much use of Gitlab issues since then because it has been easier just to keep notes locally.  I probably should change that.

A secondary reason for Gitlab is an aversion to monocultures -- I don't mind helping (in a microscopic way) to keep an alternative to Microsoft alive.  And it's not like there aren't a lot of high-profile projects on Gitlab (Gnome for one).

**Comment by parfait8566** [comment]
Pitch accent implementation discussion moved to https://gitlab.com/yamagoya/jmdictdb/-/issues/262

**Comment by JMdictProject** [comment]
With the release of the NG version recently, there's probably no reason to keep this issue open.
